File: 2608-shortest-cycle-in-a-graph/2608-shortest-cycle-in-a-graph.java
class Solution {
    public int findShortestCycle(int n, int[][] edges) {
        HashMap<Integer, HashSet<Integer>> graph = new HashMap<>();
        for(int i = 0; i < n; i++) graph.putIfAbsent(i, new HashSet<>());
        for(int[] e : edges){
            int src = e[0];
            int dst = e[1];
            graph.get(src).add(dst);
            graph.get(dst).add(src);
        }
        int ans = Integer.MAX_VALUE;
        for(int i = 0; i < n; i++){
            ArrayDeque<Integer> q = new ArrayDeque<>();
            int src = i;
            q.offer(src);

            int[] distance = new int[n];
            int[] parent = new int[n];

            Arrays.fill(distance, Integer.MAX_VALUE);
            Arrays.fill(parent, -1);

            distance[src] = 0;
            while (!q.isEmpty()){
                int size = q.size();
                while (size-- > 0){
                    int par = q.poll();
                    for(int child : graph.get(par)){
                        if(distance[child] == Integer.MAX_VALUE){
                            distance[child] = distance[par] + 1;
                            q.offer(child);
                            parent[child] = par;
                        }else if(parent[par] != child && parent[child] != par){
                            ans = Math.min(ans, distance[par] + distance[child] + 1);
                        }
                    }
                }
            }
        }
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }

}

File: 2608-shortest-cycle-in-a-graph/2608-shortest-cycle-in-a-graph.cpp
class Solution {
public:
    
    int solve(int n,vector<vector<int>> &gr) {
        int ans = INT_MAX;
 
    for (int i = 0; i < n; i++) {
 
       
        vector<int> D(n+1, INT_MAX);
 
       
        vector<int> par(n, -1);
 
        
        D[i] = 0;
        queue<int> q;
 
     
        q.push(i);
 
       
        while (!q.empty()) {
 
         
            int x = q.front();
            q.pop();
 
          
            for (int child : gr[x]) {
 
            
                if (D[child] == INT_MAX) {
 
                    D[child] = 1 + D[x];
 
                   
                    par[child] = x;
 
                  
                    q.push(child);
                }
 
              
                else if (par[x] != child and par[child] != x)
                    ans = min(ans, D[x] + D[child] + 1);
            }
        }
    }
 
   
    if (ans == INT_MAX)
        return -1;
 
    
    else
        return ans;
    }
    
    int findShortestCycle(int n, vector<vector<int>>& edges) {
        
        vector<vector<int>> gr(n+1);
        
        for(int i=0;i<edges.size();i++) {
            gr[edges[i][0]].push_back(edges[i][1]);
            gr[edges[i][1]].push_back(edges[i][0]);
        }
        
        return solve(n,gr);
        
    }
};